pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -117,88 +122,193 @@ func (n *NetworkPolicyController) deleteCNP(old interface{}) {
 	n.deleteDereferencedAddressGroups(oldInternalNP)
 }
 
+// filterPerNamespaceRuleACNPsByNSLabels gets all ClusterNetworkPolicy names that will need to be
+// re-processed if a Namespace adds or removes the input labels.
+func (n *NetworkPolicyController) filterPerNamespaceRuleACNPsByNSLabels(nsLabels labels.Set) sets.String {
+	n.internalNetworkPolicyMutex.Lock()
+	defer n.internalNetworkPolicyMutex.Unlock()
+
+	affectedPolicies := sets.String{}
+	nps, err := n.internalNetworkPolicyStore.GetByIndex(store.PerNamespaceRuleIndex, store.HasPerNamespaceRule)
+	if err != nil {
+		klog.Errorf("Error fetching internal NetworkPolicies that have per-Namespace rules: %v", err)
+		return affectedPolicies
+	}
+	for _, np := range nps {
+		internalNP := np.(*antreatypes.NetworkPolicy)
+		for _, sel := range internalNP.PerNamespaceSelectors {
+			if sel.Matches(nsLabels) {
+				affectedPolicies.Insert(internalNP.SourceRef.Name)
+				break
+			}
+		}
+	}
+	return affectedPolicies
+}
+
+// addNamespace receives Namespace ADD events and triggers all ClusterNetworkPolicies that have a
+// per-namespace rule applied to this Namespace to be re-processed.
+func (n *NetworkPolicyController) addNamespace(obj interface{}) {
+	defer n.heartbeat("addNamespace")
+	namespace := obj.(*v1.Namespace)
+	klog.V(2).Infof("Processing Namespace %s ADD event, labels: %v", namespace.Name, namespace.Labels)
+	affectedACNPs := n.filterPerNamespaceRuleACNPsByNSLabels(namespace.Labels)
+	for cnpName := range affectedACNPs {
+		if cnp, err := n.cnpLister.Get(cnpName); err == nil {
```

Comment:
Sorry I think I missed understood that comment. I thought you were suggesting we enqueue the NP which is already done. The problem is that simply enqueuing the internal NP is not enough... `syncInternalNetworkPolicy` only updates the NP span, but for Namespace update events, we need the CNP to be re-processed because the cardinality of per-NS rules might have changed because of Namespace label updates

Reply:
I'm not sure if using `updateCNP` here would introduce some problems: the function was supposed to be called serially and with continuous state changes, but now it may be called concurrently and with some unexpected intermediate state. 
For example, before the change it's only possible to receive:
```
update(A1, A2) => update(A2, A3)
```
After the change it's possible to receive:
```
update(A1, A2) => update(A3, A3) => update(A2, A3)
```
And `update(A3, A3)` may be executed at the same as others.
In a corner case, it may add a deleted CNP back after `deleteCNP` deletes it.

Reply:
I don't think update(A3, A3) executed with update(A2, A3) will cause any unexpected behaviors. If you look at `updateCNP()`, the old CNP is merely used to retrieve the key of the original internalNP, which in turn is only used to preserve span; in addition, the key of the original internalNP is the same as the new CNP, so the oldCNP spec is effectively worthless tbh... and adding a deleted CNP back does not seem possible to me as well, since before we call update we use the cnpLister to check if the CNP corresponded to the internalNP is still 'alive'.
Nevertheless, for clarity I added a new `reprocessCNP` func that does not involve update(CNP, CNP) and checks for internal NP existence before re-processing CNP.

Reply:
I think adding a deleted CNP back is possible when the DELETE event event is received and `Delete(A3)` is executed between the existence check and `update(A3, A3)` as there is no mutex to prevent it from not happining. There are many yield points in `update(A3, A3)` too that can increase the possibility. 
Adding a mutex and internal NP existence check may avoid it.

Reply:
The mutex and existence check has been added in the latest commit.